### ISSUE TITLE
Track recording doesn't record in the background (Android 8, EMUI 8)

### ISSUE THREAD

**Issue by vshcherb** [osmandapp/Osmand#5632]
I noticed a bad bug in current release: during navigation Osmand very frequently shut down when running in background. I have installed Osmand+ 3.0.4 on Huawei P10 with Android 8.0.

The issue is that during navigation after some time Osmand close instead than keep running in background and consequently the navigation stops. I do not think it is related with strict battery optimization policy of Huawei phones, because it never happen before.

I started to notice this only with last version.

**Comment by sonora** [comment]
Yes, agree, I guess we can cross-link this to issues like #5578?

**Comment by MarkusTV** [comment]
I don't know if it can be useful for investigation, anyway I noticed that when I reopen Osmand after the crash in background it shows a 5 seconds countdown message asking if continue or not the interrupted route navigation.
It means that Osmand realizes that something happen and try to resume.

**Comment by sonora** [comment]
Yes, that is a standard feature we have if OsmAnd was closed while a navigation was ongoing. It is a result of the issue, not the cause of it.

**Comment by MarkusTV** [comment]
Ok got it, thank you.
My question is if Osmand crashes itself or it is forced closed by o.s  (Android or Huawei EMUI maybe for battery optimization purposes).
I do not think so because the issue is new, appeared in this last version for first time only.

Marco

Da: Hardy <notifications@github.com>
Inviato: sabato 7 luglio 2018 22:18
A: osmandapp/Osmand <Osmand@noreply.github.com>
Cc: MarkusTV <marco.brun@outlook.com>; Manual <manual@noreply.github.com>
Oggetto: Re: [osmandapp/Osmand] OsmAnd very frequently shut down when running in background. (#5632)


Yes, that is a standard feature we have if OsmAnd was closed while a navigation was ongoing. It is a result of the issue, not the cause of it.

**Comment by sonora** [comment]
Thar's interesting, so far I had attributed it to OS behavior, but you are saying it could be an OsmAnd regression?

**Comment by MarkusTV** [comment]
No idea, I cannot say that, I see that Osmand closes but I don't know if closes itself or it is killed by the OS.

Marco
________________________________
Da: Hardy <notifications@github.com>
Inviato: domenica 8 luglio 2018 10:01
A: osmandapp/Osmand
Cc: MarkusTV; Manual
Oggetto: Re: [osmandapp/Osmand] OsmAnd very frequently shut down when running in background. (#5632)


Thar's interesting, so far I had attributed it to OS behavior, but you are saying it could be an OsmAnd regression?

**Comment by sonora** [comment]
I suppose excluding OsmAnd from any app optimization does not fix it? http://www.androidbeat.com/2017/03/how-to-stop-android-apps-killed-marshmallow-nougat/

**Comment by MarkusTV** [comment]
Unfortunately not. I tried many other settings (ignore battery optimization, keep running in background, keep running under lock screen, whitelist, put in protected app list...) but no changes.
I cannot say for sure but I suspect of Osmand because it never happen in previous versions, it is a new issue, much newer than Marshmallow and its doze battery management.

Marco
________________________________
Da: Hardy <notifications@github.com>
Inviato: domenica 8 luglio 2018 15:23
A: osmandapp/Osmand
Cc: MarkusTV; Manual
Oggetto: Re: [osmandapp/Osmand] OsmAnd very frequently shut down when running in background. (#5632)


I suppose excluding OsmAnd from any app optimization does not fix it? http://www.androidbeat.com/2017/03/how-to-stop-android-apps-killed-marshmallow-nougat/

**Comment by sonora** [comment]
Can you reliably  pinpoint with which version it started?

**Comment by MarkusTV** [comment]
I cannot say for sure but it is recent, maybe in current 3.0.4 or maximum in the previous one, but it is not older than this.

Marco
________________________________
Da: Hardy <notifications@github.com>
Inviato: domenica 8 luglio 2018 23:32
A: osmandapp/Osmand
Cc: MarkusTV; Manual
Oggetto: Re: [osmandapp/Osmand] OsmAnd very frequently shut down when running in background. (#5632)


Can you reliably pinpoint with which version it started?

**Comment by Idrive-walking** [comment]
Hi, I'm a very old user of Osmand+ (from 2012). Me too I've noticed the "supposed bug" of Osmand but I think it is also a problem in Oreo 8.0 because I tried many others applications to navigate along tracks or to record tracks while walking and the only one that is not closing while in background and screen off is (on my phone, of course) RunKeeper. I tried Orux Maps, Wikiloc, ViewRanger, Locusmap and all of them are closed by system after 5-7 minutes. I hope this can help

**Comment by MarkusTV** [comment]
Hi, it is probably a particular combination of Operating System (Android 8.0), and Phone user interface (I use Huawei EMUI).
But you are right, some apps are affected and some not. I hope that Osmand will be able to stay with best ones.

Da: Idrive-walking <notifications@github.com>
Inviato: venerdì 20 luglio 2018 13:34
A: osmandapp/Osmand <Osmand@noreply.github.com>
Cc: MarkusTV <marco.brun@outlook.com>; Manual <manual@noreply.github.com>
Oggetto: Re: [osmandapp/Osmand] OsmAnd very frequently shut down when running in background. (#5632)


Hi, I'm a very old user of Osmand+ (from 2012). Me too I've noticed the "supposed bug" of Osmand but I think it is also a problem in Oreo 8.0 because I tried many others applications to navigate along tracks or to record tracks while walking and the only one that is not closing while in background and screen off is (on my phone, of course) RunKeeper. I tried Orux Maps, Wikiloc, ViewRanger, Locusmap and all of them are closed by system after 5-7 minutes. I hope this can help

**Comment by Idrive-walking** [comment]
Hi, Marco (italiano ... anch'io),
I forgot to say I'm using a Huawei Honor 8pro and Emui 8.0.0.366

**Comment by MarkusTV** [comment]
Ciao!
Yes, the combination between Android 8.0 and Huawei EMUI 8.0 is a nightmare.
But as I said, some apps are able to avoid these problems, hope that Osmand will find a fix too.

Marco

Da: Idrive-walking <notifications@github.com>
Inviato: venerdì 20 luglio 2018 15:43
A: osmandapp/Osmand <Osmand@noreply.github.com>
Cc: MarkusTV <marco.brun@outlook.com>; Manual <manual@noreply.github.com>
Oggetto: Re: [osmandapp/Osmand] OsmAnd very frequently shut down when running in background. (#5632)


Hi, Marco (italiano ... anch'io),
I forgot to say I'm using a Huawei Honor 8pro and Emui 8.0.0.366

**Comment by Idrive-walking** [comment]
Hi, today I've made a trip with my old Samsung Note 3, Android 5.0 and Osmand+ 3.0.4 everything worked fine and voice guidance was up all the time. Hope this helps

**Comment by vshcherb** [comment]
Indeed we are expecting that some applications are using different "target-sdk" i.e. they target different Android version which triggers different execution path in Android 8.0. 

The biggest change in Android 8.0 if the process running in the background, it must have a notification otherwise it is going to be killed. With last release we targetted Android 8.0 & implemented that requirement. There are still lots of combinations in the world, apps target Android 4.* or Android 5... and run on Android 8.0, in that case it is not really clear how OS will operate. 

Unfortunately we can't lower target Android SDK cause Google Play forbids from August to lower SDK than 8.0 and from November none of the application could be published with lower target SDK than 8.0.

Currently, we are investigating if we forget to enable something or specify a certain flag. If we don't find and we don't manage to reproduce on devices we have, it will confirm a bug in Android 8.0 OS of a specific vendor. What would be really annoying that we target Android 8.0 and have most of the issues on Android 8.0 and not on Android 7 for example.

**Comment by Idrive-walking** [comment]
Thanks for the explaination. I have also a strange idea (near science
fiction): could it possible that inside Android there is a "white list" of
apps that the system don't kills? Or else is the issue connected to the way
of tracking Gps? Two apps I use that are born for sport and training
(Runkeeper and Mapmyhike) continue working in background other, born for
navigation (on-line or off-line), are killed by android (OruxMaps,
LocusMap, Osmamd, Wikilock, Viewranger etc. but not Google Map). Does this
make any sense?
Guido

Il dom 22 lug 2018, 14:00 vshcherb <notifications@github.com> ha scritto:

**Comment by stacybrock** [comment]
I recently upgraded my phone to Android 8 (Oreo) and am seeing this issue when trip recording is turned on and the phone is on battery power.

- OsmAnd 3.0.4
- Android 8.0.0
- Samsung Galaxy S7 (USA T-mobile)

Please let me know if you need more information. I'm going to try recording another trip tonight so I can provide an example GPX file.

**Comment by MarkusTV** [comment]
Hi,
I never tried if the issue happens during track recording, probably it does.
For me it happen always during navigation after 5-6 minutes in lock screen.

Marco
________________________________
Da: vshcherb <notifications@github.com>
Inviato: martedì 24 luglio 2018 20:27
A: osmandapp/Osmand
Cc: MarkusTV; Manual
Oggetto: Re: [osmandapp/Osmand] Track recording doesn't record in the background on Android 8 (#5632)


So far tested on OnePlus3T, Samsung S8, Xiaomi Mi A1 with Android 8. And track recording works fine in the background.

**Comment by matthewblain** [comment]
What does this mean:
> So far tested on OnePlus3T, Samsung S8, Xiaomi Mi A1 with Android 8. And track recording works fine in the background.

Is that with a new version of Osmand? As I noted in issue #5587 it definitely stops working under some circumstances on my Moto X4. I have not tried navigation in the background (the original issue in this bug) as I almost never do that.).

**Comment by MarkusTV** [comment]
Hi, I found a workaround for this, not nice but it works.
Android kills Osmand when in background after 5-6 minutes.
Ok then I set repeat vocal announcements every 3 minutes and turn on the screen on vocal announcements.
So at least every 3 minutes the Android countdown is reset and restart from zero.
I repeat: not a nice solution but it works.
Maybe Osmand should be able to "simulate" this situation without necessarily turn on the screen.

Marco
________________________________
Da: matthewblain <notifications@github.com>
Inviato: lunedì 30 luglio 2018 03:22
A: osmandapp/Osmand
Cc: MarkusTV; Manual
Oggetto: Re: [osmandapp/Osmand] Track recording doesn't record in the background on Android 8 (#5632)


What does this mean:

So far tested on OnePlus3T, Samsung S8, Xiaomi Mi A1 with Android 8. And track recording works fine in the background.

Is that with a new version of Osmand? As I noted in issue #5587<https://github.com/osmandapp/Osmand/issues/5587> it definitely stops working under some circumstances on my Moto X4. I have not tried navigation in the background (the original issue in this bug) as I almost never do that.).

**Comment by Idrive-walking** [comment]
hi Marco,
I've had the same idea but, for me did not worked.
But I have'nt tried switching on the screen.(me idiot) 😩
Guido

Il gio 2 ago 2018, 17:55 MarkusTV <notifications@github.com> ha scritto:

**Comment by MarkusTV** [comment]
Yes, the vocal announcement is not enough to convince Android to do not kill the process, but the screen switch on works.
Of course it is more an experiment than a solution, having redundant vocal announcements and unnecessary screen turning on is annoying and not pleasant for the battery.

Marco
________________________________
Da: Idrive-walking <notifications@github.com>
Inviato: giovedì 2 agosto 2018 19:37
A: osmandapp/Osmand
Cc: MarkusTV; Manual
Oggetto: Re: [osmandapp/Osmand] Track recording doesn't record in the background on Android 8 (#5632)

hi Marco,
I've had the same idea but, for me did not worked.
But I have'nt tried switching on the screen.(me idiot) 😩
Guido

Il gio 2 ago 2018, 17:55 MarkusTV <notifications@github.com> ha scritto:

**Comment by lehenry** [comment]
Hello,

I have seen the issue on a Huawei P20 Pro, Android 8.1, EMUI 8.1.

After disabling all battery optimisations settings for OSMAnd, the issue persisted.

The notification shows when the tracking is enabled. 
When the OSM runs on the background, and the screen is switched off, then on again, 5+ second later, the notification has disappeared.
If I open again the application, the notification shows up again, but the tracking in between didn't work.

My guess is that a clean-up of notifications cause the tracking to stop/pause.

**Comment by MRigal** [comment]
I can confirm the behaviour on my Honor 10 with Android 8 EMUI 8

**Comment by MarkusTV** [comment]
I do not know if Osmand can do something, maybe it is over their possibilities, but the issue is very serious, unbelievable that Huawei, Honor, ignore it.
Osmand, AlpineQuest, Orux, Viewranger, maybe almost all similar apps (except Google Maps of course) are killed after about 5 minutes even setting any possible protection (allow running in background, ignore battery optimization, whitelist in protected apps, etc.). All these settings are simply ignored.
This of course makes Osmand (as well as all other apps) almost useless: impossible to use navigation, impossible to record a track.
I heard that the issue is much wider, it does not concerns only navigation apps but also other apps which typically need to remain active in background, like music players, etc.
For example I read that VLC banned Huawei, they made impossible to install VLC from PalyStore in Huawei devices.
I am afraid that we are in the middle of a battle bigger than Osmand.

Marco

Da: Matthieu Rigal <notifications@github.com>
Inviato: venerdì 31 agosto 2018 10:51
A: osmandapp/Osmand <Osmand@noreply.github.com>
Cc: MarkusTV <marco.brun@outlook.com>; Manual <manual@noreply.github.com>
Oggetto: Re: [osmandapp/Osmand] Track recording doesn't record in the background (Android 8, EMUI 8) (#5632)


I can confirm the behaviour on my Honor 10 with Android 8 EMUI 8

**Comment by bsidhom** [comment]
I'm seeing this issue as well with OsmAnd 3.1.5. I'm using a Pixel 2 XL on Android 9. I've had "battery optimization" disabled for OsmAnd ever since the doze feature was introduced. I'm able to record GPX logs for a few hours before it cuts out abruptly. I only notice that it has stopped recording when I reopen the app to check on it.

**Comment by MarkusTV** [comment]
Bad news, it means it is not only Huawei and no hope to see this fixed in the future.
It means it is not a bug but a strategy to force users to do not use certain apps.

Marco
________________________________
Da: Ben Sidhom <notifications@github.com>
Inviato: martedì 4 settembre 2018 07:39
A: osmandapp/Osmand
Cc: MarkusTV; Manual
Oggetto: Re: [osmandapp/Osmand] Track recording doesn't record in the background (Android 8, EMUI 8) (#5632)


I'm seeing this issue as well with OsmAnd 3.1.5. I'm using a Pixel 2 XL on Android 9. I've had "battery optimization" disabled for OsmAnd ever since the doze feature was introduced. I'm able to record GPX logs for a few hours before it cuts out abruptly. I only notice that it has stopped recording when I reopen the app to check on it.

**Comment by MarkusTV** [comment]
Hi,
I did not try VLC, I just reported some news I heard, if VLC works fine much better.
If it is only an issue of some apps working in background with GPS I am happy, in this case I hope it will be fixed soon.

Marco

Da: Idrive-walking <notifications@github.com>
Inviato: martedì 4 settembre 2018 08:22
A: osmandapp/Osmand <Osmand@noreply.github.com>
Cc: MarkusTV <marco.brun@outlook.com>; Manual <manual@noreply.github.com>
Oggetto: Re: [osmandapp/Osmand] Track recording doesn't record in the background (Android 8, EMUI 8) (#5632)

Hi Marco,
I've installed VLC on my Honor 8pro and Android 8.0.0 without any problem, so, again, I think that the issue is on some settings of some apps working in background with GPS.

**Comment by sonora** [comment]
The Samsung XCover4 has received its update from Android 7.0.1 to Android 8.1. The previous trip logging behavior was documented in #5255.

A few tests now with 8.1. Conditions:
- Device Power Saving setting is set to OFF in the Android settings
- For what it's worth, OsmAnd has been added to the list of ""Unmonitored apps, they can use as much power as they want"
- Location service is set to Device (GPS) only
- No long term test yet, only for a few minutes each

**Manual trip recording with interval 3 sec** (i.e. OsmAnd keeps the GPS running continuously): This seems to work only as long as the device screen is on (both with OsmAnd being the app displayed in the foreground or while another app is shown). As soon as the screen turns off, no points seem logged any more, but logging is resumed as soon as the device is turned on again.

**Manual trip recording with interval 30 sec** (i.e. OsmAnd shuts down GPS in between): It looks like this works reliably only while OsmAnd is the app displayed in the foreground. As soon as another app is in the foreground, it GPS occasionally comes on and a point is being logged, but e.g. only after 90 sec or so, it is not reliable. While the device screen is turned off, it seems no points are logged at all.

In summary, reliable all-day trip logging seems impossible with OsmAnd now, I am falling back to Android 4.4 devices where this works.

**Comment by sonora** [comment]
Comparing all feedback we have here, it looks to me like we are facing two facets of this isuue:
- (1) Is that the process is actually killed by the OS under some circumstances when in the background
 -(2) Is that the process is never killed, but still its execution is somehow limited/bundled by the OS (even if whitelisted in Power settings or Power savibg is disabled), leading to e.g. the GPS being turned on only every 5 minutes even if we specify 30 seconds. Looks lile that behavior started with Android 5 or 6 and is present throuhh 8.1, but is no issue on Android 4.x

**Comment by vshcherb** [comment]
We have potential fix for Android 8.0 though it is possible to do manually in the settings. Recently we've discovered how to fix it on MIUI (Xiaomi) - we need to add to documentation:
Battery Settings
• Settings App > Battery & Performance > Manage Apps Battery Usage > Power Saving Modes 
- Off: All background data & location access of running apps will be enabled, less battery life
- On: Default Setting, more battery life
• Settings App > Battery & Performance > Manage Apps Battery Usage > Choose apps
- From there you can configure the background data & location services for each app.
• Settings App > Additional Settings > Battery & Performance > System Animations
- Hide System Animations: Speeds up device's performance

**Comment by MarkusTV** [comment]
Maybe every UI and every OS version is slightly different, I do not find anything similar in my Huawei EMUI 8.0.
I would be very happy to try this fix if you could add more details, thank you.

Marco

Da: vshcherb <notifications@github.com>
Inviato: lunedì 29 ottobre 2018 14:16
A: osmandapp/Osmand <Osmand@noreply.github.com>
Cc: MarkusTV <marco.brun@outlook.com>; Manual <manual@noreply.github.com>
Oggetto: Re: [osmandapp/Osmand] Track recording doesn't record in the background (Android 8, EMUI 8) (#5632)


We have potential fix for Android 8.0 though it is possible to do manually in the settings. Recently we've discovered how to fix it on MIUI (Xiaomi) - we need to add to documentation:
Battery Settings
• Settings App > Battery & Performance > Manage Apps Battery Usage > Power Saving Modes

  *   Off: All background data & location access of running apps will be enabled, less battery life
  *   On: Default Setting, more battery life
• Settings App > Battery & Performance > Manage Apps Battery Usage > Choose apps
  *   From there you can configure the background data & location services for each app.
• Settings App > Additional Settings > Battery & Performance > System Animations
  *   Hide System Animations: Speeds up device's performance

**Comment by sonora** [comment]
@vshcherb From my tests:

Issue (1') above, i.e. "the app is killed", may be the one which can be avoided (at least on some devices I tested) if you disable all device Power management. For recording to work then, you need to have recording interval 15 sec or smaller, i.e. keep GPS on all the time.

Issue (2) above (GPS does not fire up when app is in background or device is asleep) is not affected or fixed by Power management, meaning we have a general issue with our service trying to wake up GPS, some process in Android 7/8 devices seems to bundle this and allow it to happen only every 5 minutes, even with Power management set to off.

**Comment by sonora** [comment]
I am wondering if for Issue (2) we are suffering from this https://github.com/nextcloud/android/issues/2010 and should try the workaround mentioned in StackOverflow, calling startForeground in the service's onCreate.

**Comment by scaidermern** [comment]
If you can reproduce the issue: Does this problem also occur with other GPS tracking apps, like osmtracker? If they don't suffer from this problem then compare how they configure their GPS background service.

**Comment by st-anton** [comment]
Sorry for interfering in the discussion; I'm not a app developer, but a sad Huawei Andoid 8 user :-(
I searched in the internet and found two documents. I guess the two documents are already known:
[https://developer.android.com/guide/topics/location/](url)
[https://developer.android.com/about/versions/oreo/background-location-limits](url)
Especial the second worries me: Does it consequently mean, that Android 8 cannot handle navigation in background? In other words: Does Google navigation (and others) run in forground all the time?

**Comment by Zahnstocher** [comment]
> Especial the second worries me: Does it consequently mean, that Android 8 cannot handle navigation in background? In other words: Does Google navigation (and others) run in forground all the time?

Yes, a foreground service is needed, but this doesn't mean, that the App UI has to be visible.
OsmAnd already runs the service in foreground mode:
https://github.com/osmandapp/Osmand/blob/0cdbff3454cb3351a603d3e4120f8d661c1a749c/OsmAnd/src/net/osmand/plus/OsmandApplication.java#L839

**Comment by Bambasti** [comment]
I can confirm this bug. Happens on a Samsung Galaxy S5 Neo equipped with LineageOS 15.1 / Android 8.1. All possible power savings are disabled for OsmAnd+, even for vaguely related apps like GPS Status as well. Still any recording / navigation stops shortly after switching off the screen, GPS gets lost. On my Xiaomi Pocophone F1 that's running Android 9 I don't have these issues. Since I like to use my older Galaxy as a stand-alone navigation tool mounted on my bike this is quite a deal-breaker. Having the screen enabled at all times is not an option. OruxMaps is working fine and recording in the background as intended. So this is not an issue that's impossible to overcome.

**Comment by MarkusTV** [comment]
You say… “So this is not an issue that's impossible to overcome”: I agree, it is embarrassing how Osmand is ignoring and underestimating this problem which makes the app totally useless.


Da: Bambasti <notifications@github.com>
Inviato: domenica 11 novembre 2018 22:49
A: osmandapp/Osmand <Osmand@noreply.github.com>
Cc: MarkusTV <marco.brun@outlook.com>; Mention <mention@noreply.github.com>
Oggetto: Re: [osmandapp/Osmand] Track recording doesn't record in the background (Android 8, EMUI 8) (#5632)


I can confirm this bug. Happens on a Samsung Galaxy S5 Neo equipped with LineageOS 15.1 / Android 8.1. All possible power savings are disabled for OsmAnd+, even for vaguely related apps like GPS Status as well. Still any recording / navigation stops shortly after switching off the screen, GPS gets lost. On my Xiaomi Pocophone F1 that's running Android 9 I don't have these issues. Since I like to use my older Galaxy as a stand-alone navigation tool mounted on my bike this is quite a deal-breaker. Having the screen enabled at all times is not an option. OruxMaps is working fine and recording in the background as intended. So this is not an issue that's impossible to overcome.

**Comment by bsidhom** [comment]
I'm now able to quickly reproduce this on my Pixel 3 XL running Android 9.

Steps to reproduce:
1. Start the OsmAnd GPX logging service with a recording interval under 30 seconds (so the GPS does not sleep).
1. Turn off the screen and leave the device for a few minutes.
1. Open the camera app directly and take several photos rapidly (preferably with HDR/motion enabled to increase memory pressure).
1. Check that the OsmAnd GPX logging notification has disappeared. If not, repeat steps 2-3.

I noticed the following lines in logcat output:
```
11-19 20:10:59.842   859   859 E lowmemorykiller: /dev/memcg/apps/uid_10018/pid_3892/memory.stat open failed: No such file or directory
11-19 20:10:59.843   859   859 I lowmemorykiller: Killing 'android.process.media' (3892), uid 10018, adj 700
11-19 20:10:59.843   859   859 I lowmemorykiller:    to free 29424kB because system is under low memory pressure oom_adj 100
11-19 20:10:59.843   859   859 E lowmemorykiller: /dev/memcg/apps/uid_10092/pid_5559/memory.stat open failed: No such file or directory
11-19 20:10:59.843   859   859 I lowmemorykiller: Killing 'com.google.android.tts' (5559), uid 10092, adj 201
11-19 20:10:59.843   859   859 I lowmemorykiller:    to free 67036kB because system is under low memory pressure oom_adj 100
11-19 20:10:59.844   859   859 E lowmemorykiller: /dev/memcg/apps/uid_10167/pid_5421/memory.stat open failed: No such file or directory
11-19 20:10:59.844   859   859 I lowmemorykiller:    to free 37032kB because system is under low memory pressure oom_adj 100
11-19 20:10:59.844   859   859 E lowmemorykiller: /dev/memcg/apps/uid_10150/pid_2536/memory.stat open failed: No such file or directory
11-19 20:10:59.845   859   859 I lowmemorykiller: Killing 'net.osmand.plus' (2536), uid 10150, adj 200
11-19 20:10:59.845   859   859 I lowmemorykiller:    to free 288264kB because system is under low memory pressure oom_adj 100
11-19 20:10:59.845   859   859 I lowmemorykiller: Killing because cache 91424kB is below limit 92160kB for oom_adj 100
11-19 20:10:59.845   859   859 I lowmemorykiller:    Free memory is 13792kB above reserved
11-19 20:10:59.845   859   859 I lowmemorykiller: Reclaimed enough memory (pages to free=77192, pages freed=105439)
```

And below that:
```
11-19 20:11:00.278  1257  1643 W InputDispatcher: channel '2960ac net.osmand.plus/net.osmand.plus.activities.MapActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
11-19 20:11:00.279  1257  1643 E InputDispatcher: channel '2960ac net.osmand.plus/net.osmand.plus.activities.MapActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
11-19 20:11:00.299  1257  1625 I GnssLocationProvider: WakeLock acquired by sendMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@ac12a1a)
11-19 20:11:00.299  1257  1271 I GnssLocationProvider: WakeLock released by handleMessage(REPORT_SV_STATUS, 0, com.android.server.location.GnssLocationProvider$SvStatusInfo@ac12a1a)
11-19 20:11:00.311  1257  3397 I GnssLocationProvider: WakeLock acquired by sendMessage(SET_REQUEST, 0, com.android.server.location.GnssLocationProvider$GpsRequest@92f094b)
11-19 20:11:00.311  1257  1275 W libprocessgroup: kill(-2536, 9) failed: No such process
11-19 20:11:00.311  1257  3493 I ActivityManager: Process net.osmand.plus (pid 2536) has died: prcp FGS 
11-19 20:11:00.311  1257  3516 I WindowManager: WIN DEATH: Window{2960ac u0 net.osmand.plus/net.osmand.plus.activities.MapActivity}
11-19 20:11:00.311  1257  3516 W InputDispatcher: Attempted to unregister already unregistered input channel '2960ac net.osmand.plus/net.osmand.plus.activities.MapActivity (server)'
11-19 20:11:00.312  1257  3493 W ActivityManager: Scheduling restart of crashed service net.osmand.plus/.NavigationService in 1000330ms
```

That last log entry makes it seem like the service should be restarted in about 15 minutes. However, it took about 30 minutes to come up on its own in my test.

So in my case the issue appears to be memory pressure. Is the (foreground) service started with the [`START_STICKY`](https://developer.android.com/reference/android/app/Service#START_STICKY) flag? If not, this could shorten recovery times. I'm not sure there's anything that can be done to prioritize OsmAnd over other long-running services to prevent eviction.

See https://stackoverflow.com/a/9441795 and https://android-developers.googleblog.com/2010/02/service-api-changes-starting-with.html for more details on `START_STICKY`.

**Comment by bsidhom** [comment]
If `START_STICKY` is not robust enough (it seems like there may be a large delay in getting the service restarted), it's possible that [`JobScheduler`](https://developer.android.com/reference/android/app/job/JobScheduler) with a tight deadline can help bootstrap a restart.

**Comment by bsidhom** [comment]
Just confirmed that [`START_REDELIVER_INTENT`](https://developer.android.com/reference/android/app/Service#START_REDELIVER_INTENT) is being used instead of `START_STICKY` at https://github.com/osmandapp/Osmand/blob/cdf49f10f616d17214767694f3c7b5333c81201a/OsmAnd/src/net/osmand/plus/NavigationService.java#L170. The [documentation](https://developer.android.com/reference/android/app/Service#service-lifecycle) is not clear on the nuances of when either should be used but _does_ imply that `START_STICY` should be used for services that are explicitly started and stopped (which fits the bill here).

**Comment by bsidhom** [comment]
Another option might be to use the broadcast hack explained here: https://fabcirablog.weebly.com/blog/creating-a-never-ending-background-service-in-android. My main concern with that approach is that I'm not sure whether broadcasts might be swallowed. That approach depends on the receiver running in a different process than the sender. There appears to be a race between the broadcast, process termination, and broadcast receipt. If the broadcast is received by the process that is sending termination, I suspect it will be lost.

**Comment by Bambasti** [comment]
Tested another device I have. On a Motorola G (2nd gen.) on LineageOS 14.1 / Android 7.1.2 it's working in the background as well, like on Android 9. So this issue seems to be about Android 8/8.1 only.

**Comment by gormo-osm** [comment]
The issue started for me with the upgrade of Android or EMUI to 8. Now both are at 8. Device is a P8 Lite 2017. GPX tracking stops ca. 5 Minutes after screen turnoff.

**Comment by st-anton** [comment]
Last week I tried the background tracking, and found, that it work again as it should do. I'm happy with it again as say thanks to the developers.

**Comment by sonora** [comment]
Looks like this is now our master issue for this still very annoying problem. For me the current status is as follows: (Android 8.1, all device power management turned off, for good measure also exempted OsmAnd from all app optimizations. Location method is set to GPS only.).
- For GPX recording intervals up to and including 15s, for which we do not turn off GPS between recording points, track recording now works as expected, even while the screen is off.
- But for recording intervals of 30+ seconds, where we turn off GPS to conserve power, we still seem to encounter a location limitation while the screen is off: For e.g. 30 sec. I receive a recorded point only every 5 minutes.

I have found this https://hackernoon.com/android-location-tracking-with-a-service-80940218f561 and am aware that we are now using a foreground service in OsmAnd, but perhaps the note and code about the notification channel is helpful, not sure.

In any case, we now have a clear distinction between whe it works and when not, I observe the issue only in the cases where we shut down GPS between measurements. So it should be fixable.

This is one of the major bugs we now have in OsmAnd, it definitely deserves to be in a milestone with adequate priority!

**Comment by sonora** [comment]
Looking at our code, it seems that we use AlarmManager to trigger recording a point for recording intervals >=30sec, see https://github.com/osmandapp/Osmand/blob/0cdbff3454cb3351a603d3e4120f8d661c1a749c/OsmAnd/src/net/osmand/plus/NavigationService.java#L104.

As per Android documentation:

> Note: Beginning with API 19 (Build.VERSION_CODES.KITKAT) alarm delivery is inexact: the OS will shift alarms in order to minimize wakeups and battery use. There are new APIs to support applications which need strict delivery guarantees; see setWindow(int, long, long, android.app.PendingIntent) and setExact(int, long, android.app.PendingIntent). Applications whose targetSdkVersion is earlier than API 19 will continue to see the previous behavior in which all alarms are delivered exactly when requested.

Maybe we should use one of the other recommended ways to schedule recording points?

**Comment by vshcherb** [comment]
Yes we need to be more than API 19 and try the approaches you mentionned. I will put it in 3.4 release so we don't forget about it.

**Comment by sonora** [comment]
More reading on the subject: https://stackoverflow.com/questions/33110246/setexactandallowwhileidle-is-not-exact-as-of-developer-reference. Would have thought we simply replace our 2 occurrences of `setRepeating` by `setExactAndAllowWhileIdle`, and re-schedule in `onReceive`, but it seems unclear if this works reliably...

**Comment by sonora** [comment]
I have created a pull request laying the foundation by using the correct `setExact / setExactAndAllowWhileIdle` for higher Build levels, please review and merge as step 1.

I guess step 2 could be something along the lines of https://medium.com/@sauge16/how-to-handle-alarm-in-all-android-version-e7aca16ae885.

**Comment by sonora** [comment]
My test results after this commit:
 - (1) While OsmAnd is the foreground app: On all devices I have access to (Android 4.1 - 8.1), logging now works as expected, both for short-interval (<=15sec) and long-interval (>=30sec) logging intervals.
- (2) While OsmAnd is not the foreground app: The same is true, as long as the device screen is on
- (3) All logging also works as expected while the device screen is off, up to and including KitKat devices
- (4) The major issue is still the device screen being off for Android 6+ devices: It looks like on most devices the points recorded soon drop to a 3 minute, 5 minute, or even 10 minute interval, depending on device.
- (5) Side note (but no new observation): In addition, particularly for device-off cases, some scheduled measurement points are missing in the resulting GPX file. Maybe this is due to a GPS fix not being established in time prior to the device stopping to try.
- (6) Side effect: looks like at least on some Android 8+ devices, the Android notification is not updated continuously: While the device has in fact recorded points for a long time (albeit e.g. at 5 minute intervals), the Android notification may not reflect this correctly after the screen is turned back on, i.e. it may be far behind indicating the actual track which has beenwritten. Keep this effect in mind when testing, inspect the actual GPX file, do not rely on the notification.

The major functional issue is clearly (4), i.e. the device not repeatedly waking up at exact times in doze mode for new Android versions, still not sure how to solve ...

**Comment by sonora** [comment]
The only (I hope temporary) workaround I can think of for now is that for logging
 - for `Build.VERSION.SDK_INT >= 19`, and
 - for  logging intervals `3ß sec <= x < 5 min`
we resort to using `serviceOffInterval = 0`, i.e. keep the GPS on continuously as we do today for intervals <= 15 sec?

We have hard-coded the 30 sec in `AppInitializer.java`, `OsmAndMonitoring.java`, `NavigationService.java`, and perhaps elsewhere. Can provide a pull request for verification if desired.

**Comment by sonora** [comment]
I have created pull request "Issue 5632 Workaround". Commenting out the first if block in `navigationServiceGpsInterval()` will re-establish our old (and desired) default behavior.

**Comment by sonora** [comment]
First test results on one device (Android 4.4.2):
- The pull request works as expected
- Battery use is about 5% per hour on that device, not nearly as bad as I would have thought
- The issue of occasionally missing points (number (5) further up) has entirely disappeared, and the accuracy of recorded tracks has significantly improved, due to the fact that a GPS fix is kept if possible.

From what I see now we should go with this solution at this point.

**Comment by vshcherb** [comment]
Great work! @sonora

**Comment by sonora** [comment]
@vshcherb  Much appreciated. Thx, Victor   ;)

**Comment by sonora** [comment]
Let me apply the workaround to only Android 6+ devices for now, see https://github.com/osmandapp/Osmand/commit/27cb7bc5615cd3703f923366e853164aa767a42a.

Should anybody have any Android 4.4.x or Android 5.x devices having issues with background logging for =>30sec intervals, we can include those SDKs, too. But for devices I own it seems not necessary.

**Comment by sonora** [comment]
To consolidate our documentation on findings and strategies regarding background behavior, I have created a brief Tech Article on **Background Track Recording - Strategies and Known Issues** here: https://osmand.net/help-online/technical-articles#BackgroundRecording.

**Comment by e-pirate** [comment]
Heh, I'm here from #7083 . GPX recording just simply stops on my BlackView BV9500 if a lock the phone with OsmAnd+ not in foreground making the app almost useless.

**Comment by scaidermern** [comment]
Track recording stopped on my device (Xiaomi A1) after about 1.5 hours with OsmAnd being in the foreground but display off.

OsmAnd 3.4.0 on Android 8.1 (LineageOS) . Roughly the same behavior as in the previous OsmAnd version.

**Comment by sonora** [comment]
@e-pirate Please try OsmAnd 3.4, from what I can see on my devices that issue should be fixed now.

@scaidermern That one sounds like the Android 8.x Power management killed OsmAnd after some time. Have you tried white-listing OsmAnd in the Android Power settings? (Or, for the sake of testing, just disable Android's Power saving entirely on the device?)

**Comment by scaidermern** [comment]
> @scaidermern That one sounds like the Android 8.x Power management killed OsmAnd after some time. Have you tried white-listing OsmAnd in the Android Power settings? (Or, for the sake of testing, just disable Android's Power saving entirely on the device?)

So far I didn't white-list OsmAnd. I will do this, thanks.

**Comment by heitorPB** [comment]
With OSMand on foreground and Trip Recording plugin recording a track, if I lock the screen OSMand stops recording the track. But If I unlock the screen again, OSMand adds another point to the track. Maybe the problem is I/O when screen is locked? It doesn't look like the system lost the GNSS fix.

Osmand 3.3.8, XIAOMI Redmi Note 4x, MIUI Global 10.2 with Android 7.0.

**Comment by sonora** [comment]
I have summarized here: https://osmand.net/help-online/faq#trip_recording_stops

**Comment by topas-rec** [comment]
Maybe @sonora you can add the following to the Osmand help pages:

"Lock the application in the application switcher (list active applications - locate running Osmand app - find the lock command and lock the app)

I experienced that the Osmand app got stopped in background during navigation on highway and also during track recording on bike on a OnePlus 5t with unmodified vendor OS which runs Android 9.
I then tried the suggestions in the FAQ without success.
Since a few rides (only track recording on bike, rare navigation in car) the issue seems solved by using the lock option if an app.

Perhaps one who has this issue can also try this suggestion.

**Comment by topas-rec** [comment]
> Maybe @sonora you can add the following to the Osmand help pages:
> 
> "Lock the application in the application switcher (list active applications - locate running Osmand app - find the lock command and lock the app)"
> 
> I experienced that the Osmand app got stopped in background during navigation on highway and also during track recording on bike on a OnePlus 5t with unmodified vendor OS which runs Android 9.
> I then tried the suggestions in the FAQ without success.
> Since a few rides (only track recording on bike, rare navigation in car) the issue seems solved by using the lock option if an app.
> 
> Perhaps one who has this issue can also try this suggestion.

With OnePlus update to Android 10 this does not work anymore.

**Comment by ioctl-user** [comment]
Please see patch prototype here: https://github.com/osmandapp/Osmand/issues/9097

It fix big-timeout-wake-up-problem.
